From ecedbae33c05606a91652cb3e50553bf7e962aa2 Mon Sep 17 00:00:00 2001
From: Alessandro Decina <alessandro.decina@collabora.co.uk>
Date: Mon, 9 May 2011 10:27:44 +0200
Subject: [PATCH 12/19] v4l2sink: handle GST_EVENT_LIVE_FLUSH

Unblock buffer_alloc when GST_EVENT_LIVE_FLUSH is received. Fixes deadlocks with
live sources allocating from v4l2sink's pool.
---
 sys/v4l2/gstv4l2bufferpool.c |    2 +-
 sys/v4l2/gstv4l2bufferpool.h |    2 ++
 sys/v4l2/gstv4l2sink.c       |    5 ++++-
 3 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/sys/v4l2/gstv4l2bufferpool.c b/sys/v4l2/gstv4l2bufferpool.c
index e976ec5..e9b14c5 100644
--- a/sys/v4l2/gstv4l2bufferpool.c
+++ b/sys/v4l2/gstv4l2bufferpool.c
@@ -483,7 +483,7 @@ gst_v4l2_buffer_pool_get (GstV4l2BufferPool * pool, gboolean blocking)
     buf = g_async_queue_try_pop (pool->avail_buffers);
   }
 
-  if (buf) {
+  if (buf && buf != GST_V4L2_BUFFER_SENTINEL) {
     GST_V4L2_BUFFER_POOL_LOCK (pool);
     GST_BUFFER_SIZE (buf) = buf->vbuffer.length;
     GST_BUFFER_FLAG_UNSET (buf, 0xffffffff);
diff --git a/sys/v4l2/gstv4l2bufferpool.h b/sys/v4l2/gstv4l2bufferpool.h
index caad9ac..e866650 100644
--- a/sys/v4l2/gstv4l2bufferpool.h
+++ b/sys/v4l2/gstv4l2bufferpool.h
@@ -43,6 +43,8 @@ GType gst_v4l2_buffer_pool_get_type (void);
 #define GST_IS_V4L2_BUFFER_POOL(obj) (G_TYPE_CHECK_INSTANCE_TYPE ((obj), GST_TYPE_V4L2_BUFFER_POOL))
 #define GST_V4L2_BUFFER_POOL(obj) (G_TYPE_CHECK_INSTANCE_CAST ((obj), GST_TYPE_V4L2_BUFFER_POOL, GstV4l2BufferPool))
 
+#define GST_V4L2_BUFFER_SENTINEL GINT_TO_POINTER(42)
+
 
 
 typedef struct _GstV4l2BufferPool GstV4l2BufferPool;
diff --git a/sys/v4l2/gstv4l2sink.c b/sys/v4l2/gstv4l2sink.c
index 933ba87..146803c 100644
--- a/sys/v4l2/gstv4l2sink.c
+++ b/sys/v4l2/gstv4l2sink.c
@@ -1029,7 +1029,7 @@ gst_v4l2sink_buffer_alloc (GstBaseSink * bsink, guint64 offset, guint size,
 
     v4l2buf = gst_v4l2_buffer_pool_get (v4l2sink->pool, TRUE);
 
-    if (G_LIKELY (v4l2buf)) {
+    if (G_LIKELY (v4l2buf && v4l2buf != GST_V4L2_BUFFER_SENTINEL)) {
       GST_DEBUG_OBJECT (v4l2sink, "allocated buffer: %p", v4l2buf);
       *buf = GST_BUFFER (v4l2buf);
       return GST_FLOW_OK;
@@ -1076,6 +1076,9 @@ gst_v4l2sink_event (GstBaseSink * bsink, GstEvent * event)
       gst_v4l2sink_sync_crop_fields (v4l2sink);
       return TRUE;
     }
+    case GST_EVENT_LIVE_FLUSH:
+      g_async_queue_push (v4l2sink->pool->avail_buffers, GST_V4L2_BUFFER_SENTINEL);
+      return TRUE;
     default:{
       if (GST_BASE_SINK_CLASS (parent_class)->event) {
         return GST_BASE_SINK_CLASS (parent_class)->event (bsink, event);
-- 
1.7.1

